x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -280,33 +303,142 @@ public void snapshotShard(IndexShard shard, Store store, SnapshotId snapshotId,
     public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version version, IndexId indexId, ShardId shardId,
                              RecoveryState recoveryState) {
         // TODO: Add timeouts to network calls / the restore process.
+        createEmptyStore(indexShard, shardId);
+
+        final Map<String, String> ccrMetaData = indexShard.indexSettings().getIndexMetaData().getCustomData(Ccr.CCR_CUSTOM_METADATA_KEY);
+        final String leaderIndexName = ccrMetaData.get(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_NAME_KEY);
+        final String leaderUUID = ccrMetaData.get(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_UUID_KEY);
+        final Index leaderIndex = new Index(leaderIndexName, leaderUUID);
+        final ShardId leaderShardId = new ShardId(leaderIndex, shardId.getId());
+
+        final Client remoteClient = getRemoteClusterClient();
+
+        final String retentionLeaseId =
+                retentionLeaseId(localClusterName, indexShard.shardId().getIndex(), remoteClusterAlias, leaderIndex);
+
+        acquireRetentionLeaseOnLeader(shardId, retentionLeaseId, leaderShardId, remoteClient);
+
+        // schedule renewals to run during the restore
+        final Scheduler.Cancellable renewable = threadPool.scheduleWithFixedDelay(
+                () -> {
+                    logger.trace("{} background renewal of retention lease [{}] during restore", shardId, retentionLeaseId);
+                    final ThreadContext threadContext = threadPool.getThreadContext();
+                    try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                        // we have to execute under the system context so that if security is enabled the renewal is authorized
+                        threadContext.markAsSystemContext();
+                        renewRetentionLease(leaderShardId, retentionLeaseId, remoteClient);
```

Comment:
Should we use a non-blocking call instead here?

Reply:
Good catch.

Reply:
@dnhatn I pushed 6fc440b50bae12053cd1051a56b49ab772549d93. Please let me know your thoughts.

Reply:
Looks good :).